TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs

```diff
@@ -986,13 +986,196 @@ private Node ResolveProcedure(Node node, SectionNode sectionNode, SymbolReferenc
+                {
+                    Procedure currentProcedure = this.CurrentProgramCfgBuilder.AllProcedures[currentProcedureNumber];
+                    currentProcedure.AccumulateSentencesThrough(null, throughProcedure, out var lastProcedure);
+                    currentProcedureNumber = lastProcedure.Number + 1;
+
+                    addToRelation(currentProcedure, callee);
+                }
+            }
+            else
+            {
+                addToRelation(calleeProcedure, callee);
+            }
+            callRelationDomain.Add(caller);
+
+            void addToRelation(Procedure _calleeProcedure, PerformProcedure _callee)
+            {
+                if (!callRelation.TryGetValue(caller, out var callees))
+                {
+                    callees = new Tuple<PerformProcedure, List<Node>, HashSet<Procedure>>(performCaller, new List<Node>(), new HashSet<Procedure>());
+                    callRelation[caller] = callees;
+                }
+                if (!callees.Item3.Contains(_calleeProcedure))
+                {
+                    callees.Item3.Add(_calleeProcedure);                    
+                }
+                if (!callees.Item2.Contains(_callee))
+                {
+                    callees.Item2.Add(_callee);
+                }
+                callRelationDomain.Add(_calleeProcedure);
+            }
+        }
+
+        /// <summary>
+        /// Compute the transitive closure of the CALL relation.
+        /// We browse the the graph represented using the Call Relation, then we found during the visit cycles.
+        /// In the same time we compute the call chain.
+        /// </summary>
+        /// <param name="callRelation">CALL Relation</param>
+        /// <param name="callRelationDomain">The domain of the call relation</param>
+        /// <param name="reportedProcedures">Already reported recursif procedures</param>
+        private void TransitiveClosureCallRelation(Dictionary<Procedure, Tuple<PerformProcedure, List<Node>, HashSet<Procedure>>> callRelation, HashSet<Procedure> callRelationDomain,
+            HashSet<PerformProcedure> reportedProcedures)
+        {
+            Stack <Procedure> S = new Stack<Procedure>();
+            Dictionary<Procedure, int> N = new Dictionary<Procedure, int>();
+
+            foreach(var a in callRelationDomain)
+            {
+                var na = Visited(a);
+                if (na == 0)
+                    dfs(a);
+            }
+            // Report all cyclic procedure.
+            foreach (var item in callRelation)
+            {
+                if (!reportedProcedures.Contains(item.Value.Item1))
+                {
+                    if (item.Value.Item3.Contains(item.Key))
+                    {
+                        this.Cfg.AddRecursivePerform(item.Value.Item1, item.Value.Item2);
+                    }
+                }
+            }
+
+            void dfs(Procedure a)
```

Comment:
Ok this algorithm is not my invention, it is based on J. Eve, R. Kurki-Suonio transitive closure calculation algorithm, you can google on it, there are  many adaptation. I implemented also an adaptation for gathering perform chain path.

Reply:
Do we need it ? What about this much simpler algorithm ?

```csharp
        public static Dictionary<N, HashSet<N>> TransitiveClosure<N>(Dictionary<N, HashSet<N>> graph)
        {
            var transitiveClosure = new Dictionary<N, HashSet<N>>();
            var root = graph.FirstOrDefault().Key;
            if (root != null)
            {
                ComputeDirectAndIndirectSuccessors(root);
            }
            return transitiveClosure;

            HashSet<N> ComputeDirectAndIndirectSuccessors(N node)
            {
                if (transitiveClosure.TryGetValue(node, out var result))
                {
                    return result;
                }

                var directSuccessors = graph[node];
                result = new HashSet<N>(directSuccessors);
                transitiveClosure.Add(node, result);

                foreach (var directSuccessor in directSuccessors)
                {
                    foreach (var indirectSuccessor in ComputeDirectAndIndirectSuccessors(directSuccessor))
                    {
                        result.Add(indirectSuccessor);
                    }
                }

                return result;
            }
        }
```

Reply:
Ok here is the url of the Eve paper: http://i.stanford.edu/pub/cstr/reports/cs/tr/75/508/CS-TR-75-508.pdf
It seems to me that the algorithm that you suggest is in the best case in O(n^2) and in the worst case in O(n^4) because it can implies 4 foreach loops.
In the paper it is say that computing transitive closure is in the best case be in O(n^2) and in the worst case in O(n^3), and it is said in the paper that the algorithm in the paper usually works in O(n) to O(n^2) where n is the number of vertices.

And I don't see in your algorithm where to compute perform call path chain.

Reply:
Yes it was an attempt at writing a simpler algorithm to help understanding the transitive closure. Anyway my code does not work properly for cyclic graphs so we can forget about it.

However the main question still remains, do we need to compute the full closure ? Our goal is to gather cyclic chains but not necessarily building the transitive closure of the perform call graph.

Reply:
My answer is yes we need to compute the full transitive closure, it is this computation that discovers Strongly connected components (that is to say cyclic chains).